client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java

```diff
@@ -671,8 +676,18 @@ private SendResult sendDefaultImpl(
     private TopicPublishInfo tryToFindTopicPublishInfo(final String topic) {
         TopicPublishInfo topicPublishInfo = this.topicPublishInfoTable.get(topic);
         if (null == topicPublishInfo || !topicPublishInfo.ok()) {
-            this.topicPublishInfoTable.putIfAbsent(topic, new TopicPublishInfo());
-            this.mQClientFactory.updateTopicRouteInfoFromNameServer(topic);
+            try {
+                if(lockUpdatePublishInfo.tryLock(LOCK_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS)){
```

Comment:
IMO, add a new lock before lockNamesrv.tryLock do not make effect. We can check topic route info exists or not again after lockNamesrv.tryLock success to avoid redundant getRouteInfoByTopic.

Reply:
> IMO, add a new lock before lockNamesrv.tryLock do not make effect. We can check topic route info exists or not again after lockNamesrv.tryLock success to avoid redundant getRouteInfoByTopic.

Thank you very much for your reply, but there is such a scenario, the producer during startup, will be intervals of 30 s updated routing information, if add the calibration topicRouter exists within updateTopicRouteInfoFromNameServer method, It may cause topicRouter not to update in time, so I tried to put the lock on a higher level method

Reply:
We can add a parameter isEmpty2FetchFromNamesrv in mQClientFactory.updateTopicRouteInfoFromNameServer method. if isEmpty2FetchFromNamesrv is true, and double-check topic route info is also empty after tryLock, then update from namesrv. if isEmpty2FetchFromNamesrv is false, and double-check topic route info is not empty after tryLock, then do not update from namesrv. MQClientInstance's scheduled task will update topic route info regually.

Reply:
It's a good idea, I'll try to optimize it